pkg/controller/master/mcmsettings/fleetcontroller.go

```diff
@@ -0,0 +1,141 @@
+package mcmsettings
+
+import (
+	"encoding/base64"
+	"fmt"
+
+	mgmtv3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/wrangler/pkg/relatedresource"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	defaultFleetControllerConfigMapName      = "fleet-controller"
+	defaultFleetControllerConfigMapNamespace = "cattle-fleet-system"
+	rancherInternalCASetting                 = "internal-cacerts"
+	rancherInternalServerURLSetting          = "internal-server-url"
+	apiServerURLKey                          = "apiServerURL"
+	apiServerCAKey                           = "apiServerCA"
+)
+
+// watchFleetControllerConfigMap will watch the configmap for fleet-controller setting and requeue internal-server-url and internal-cacerts to trigger reconcile of associated settings to ensure correct values are applied
+func (h *mcmSettingsHandler) watchFleetControllerConfigMap(_, _ string, obj runtime.Object) ([]relatedresource.Key, error) {
+	if cm, ok := obj.(*corev1.ConfigMap); ok {
+		if cm.Name == defaultFleetControllerConfigMapName && cm.Namespace == defaultFleetControllerConfigMapNamespace {
+			return []relatedresource.Key{
+				{
+					Name: rancherInternalCASetting,
+				},
+				{
+					Name: rancherInternalServerURLSetting,
+				},
+			}, nil
+		}
+	}
+	return nil, nil
+}
+
+func (h *mcmSettingsHandler) watchInternalEndpointSettings(_ string, setting *mgmtv3.Setting) (*mgmtv3.Setting, error) {
+	if setting == nil || setting.DeletionTimestamp != nil || (setting.Name != rancherInternalCASetting && setting.Name != rancherInternalServerURLSetting) {
+		return setting, nil
+	}
+
+	logrus.Debugf("reconcilling fleet-controller config for setting %s", setting.Name)
+
+	fleetControllerConfig, err := h.configMapCache.Get(defaultFleetControllerConfigMapNamespace, defaultFleetControllerConfigMapName)
+	if err != nil {
+		return setting, fmt.Errorf("error fetching fleet-controller configmap: %v", err)
+	}
+
+	if setting.Name == rancherInternalServerURLSetting && len(setting.Value) != 0 {
+		return h.reconcileInternalServerURLSetting(setting, fleetControllerConfig)
+	}
+
+	if setting.Name == rancherInternalCASetting && len(setting.Value) != 0 {
+		return h.reconcileInternalCASetting(setting, fleetControllerConfig)
+	}
+
+	return setting, nil
+}
+
+func (h *mcmSettingsHandler) reconcileInternalCASetting(setting *mgmtv3.Setting, fleetControllerConfig *corev1.ConfigMap) (*mgmtv3.Setting, error) {
+
+	encodedCA := base64.StdEncoding.EncodeToString([]byte(setting.Value))
+	currentVal, err := getConfigMayKey(fleetControllerConfig, apiServerCAKey)
+	if err != nil {
+		return setting, err
+	}
+
+	var patchedConfigMap *corev1.ConfigMap
+	if currentVal != encodedCA {
+		patchedConfigMap, err = patchConfigMap(fleetControllerConfig, apiServerCAKey, encodedCA)
+		if err != nil {
+			return setting, err
+		}
+
+		_, err = h.configMapClient.Update(patchedConfigMap)
```

Comment:
Please double check:

I did some manual tests again `fleet-controller` configmap, it seems that the change to this configmap will not cause helm chart `fleet-controller` to be re-deployed. 

If we delete the deployment of `fleet-controller` and update any of rancher below settings, the `fleet-controller` will be re-deployed, the new configmap is used then.
```
	if setting.Name != settings.ServerURL.Name &&
		setting.Name != settings.CACerts.Name &&
		setting.Name != settings.SystemDefaultRegistry.Name {
		return setting, nil
	}
```

The related code is:

https://github.com/rancher/rancher/blob/release/v2.7.6/pkg/controllers/dashboard/fleetcharts/controller.go#L60

Reply:
The aim of the change is to not redeploy the fleet-controller. Rancher manages the fleet controller update as part of the upgrade process. Fleet controller then manages the upgrade of downstream clusters fleet-agent. This change ensures that fleet-agent is redeployed by the fleet-controller.

Currently the fleet-controller never reconciles the cluster objects because it cannot find the `apiServerURL` defined in the config map.

The change ensures this is set, which allows the fleet controller to reconcile the imported cluster, and subsequently redeploy the fleet-agent.

https://github.com/rancher/fleet/blob/v0.7.0/pkg/controllers/cluster/import.go#L235

Reply:
Sounds reasonable.

Should the upstream Rancher add this enhancement formally later ? From the view of Harvester, this workaround has potential conflicts with Rancher & Fleet-controller in case their logic changes in the future.

Reply:
we do have a related issue for this: https://github.com/rancher/rancher/issues/41848